### PR TITLE
🎉 add ability to send success notifications on webhook

### LINKED ISSUE #4519: Allow configuring a sync_success webhook
## Tell us about the problem you're trying to solve
When a sync succeeds I want to receive a notification to a webhook, same as https://github.com/airbytehq/airbyte/pull/2706 but for success notifications.

### PR DESCRIPTION
**Pull request by prasrvenkat**
## What
Ability to get notified on successful sync as well. This makes the success/failure notifications configurable with toggle on the webhook settings page. This is based on https://github.com/airbytehq/airbyte/pull/2706 and will fix https://github.com/airbytehq/airbyte/issues/4519 

## How
* Adds two flags `sendOnSuccess` and `sendOnFailure` to indicate whether to send notification on success and/or failure. 
* Default value is `true` for `sendOnFailure` and `false` for `sendOnSuccess` to keep it same as current behavior.
* Ability in front-end to toggle both on/off. I am not that familiar with frontend in general, but tried to get a working solution, which works with current state.
  * The toggle is not auto-save, its tied to the current `formik` form so reflects only after save changes.
  * Test will send different notifications for success and failure
* The call site for successful notification is in `JobSubmitter` after finishing something that is a success. Let me know if this not the right place. Ideally, I thought notification will be downstream option after tracking but there was a notifier already available there so re-used it.
* Migration test was failing so I had to add one, let me know if that is fine, especially the version

image

## Recommended reading order
Common
1. `config.yaml` and `SlackNotificationConfiguration.yaml` for reviewing the new boolean toggles.

Backend
1. `NotificationClient.java` to see methods to notify success and failure separately.
2. `JobNotifier.java` to see handling success/failure job notifications properly, especially the updated time handling.
3. `JobSubmitter.java` to see call-site for invoking `JobNotifier` on successful job completion.
4. `Migrations` and the new migration added.

Frontend
1. `WehbhookForm` changes
2. Other files.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java [hunk 1]
```diff
@@ -188,9 +188,10 @@ public WorkspaceRead updateWorkspace(final WorkspaceUpdate workspaceUpdate) thro
   public NotificationRead tryNotification(final Notification notification) {
     try {
       final NotificationClient notificationClient = NotificationClient.createNotificationClient(NotificationConverter.toConfig(notification));
-      final String message = String.format("Hello World! This is a test from Airbyte to try %s notification settings",
-          notification.getNotificationType());
-      if (notificationClient.notify(message)) {
+      final String messageFormat = "Hello World! This is a test from Airbyte to try %s notification settings for sync %s";
+      final boolean failureNotified = notificationClient.notifyFailure(String.format(messageFormat, notification.getNotificationType(), "failures"));
+      final boolean successNotified = notificationClient.notifySuccess(String.format(messageFormat, notification.getNotificationType(), "successes"));
+      if (failureNotified || successNotified) {
```
Comment (ChristopheDuong) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:194]: Do we really need to send both success and failure (two notifications) as tests to try if the webhook works?

File: airbyte-webapp/src/locales/en.json [hunk 2]
```diff
@@ -336,10 +336,12 @@
...
   "settings.accountSettings": "Account Settings",
   "settings.changeSaved": "change saved!",
-  "settings.webhook": "Connection failure Webhook",
-  "settings.webhookTitle": "Connection failure Webhook URL",
+  "settings.webhook": "Connection status Webhook",
+  "settings.webhookTitle": "Connection status Webhook URL",
   "settings.webhookDescriprion": "Get notifications the way you want (for instance, on Slack, or other).",
   "settings.webhookTestText": "Testing the Webhook will send a “Hello World”. ",
+  "settings.sendOnSuccess": "Send success sync notifications.",
+  "settings.sendOnFailure": "Send failure sync notifications.",
```
Comment (ChristopheDuong) [inline airbyte-webapp/src/locales/en.json:344]:
```suggestion
  "settings.sendOnSuccess": "Send notifications when sync succeeds.",
  "settings.sendOnFailure": "Send notifications when sync fails.",
```

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java (hunk 1 above)
Comment (prasrvenkat) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WorkspacesHandler.java:194]: I do not have strong preference. Only reasoning I had was the flags control success and/or failure notifications so it seemed right to send both if both are enabled and send one or other if only one of them (or none if none of them) is enabled.

If you feel strongly, I can drop one. Let me know.

**Review by cgardens — COMMENTED** [review]
@prasrvenkat thank you for this PR! We really appreciate you digging into so many pieces of our system to put this together. I'm definitely happy to work with you to get this merged! 

Here's the plan.
* I left a few comments throughout the PR. LMK if anything doesn't make sense. I think there is only one significant change that I am suggesting around the data model.
* I tried to address some of the questions you left in the PR body below. Again, lmk if anything doesn't make sense!
* After we iterate one more time, I'll loop in our FE guy to help out with the refreshing issue you mentioned.

> Ability in front-end to toggle both on/off. I am not that familiar with frontend in general, but tried to get a working solution.
The toggle is not auto-save, its tied to the current formik form so reflects only after save changes.

@Jamakase can you take a look at the FE change in this PR and leave any feedback please?

> The call site for successful notification is in JobSubmitter after finishing something that is a success. Let me know if this not the right place. Ideally, I thought notification will be downstream option after tracking but there was a notifier already available there so re-used it.

I agree with your assessment that this is not the ideal place. We want to change it so that the notifier happens from within a Temporal Activity. That being said, I think what you have done for now is a good compromise and we can move the call site in the future. So let's stick with what you have!

> Migration test was failing so I had to add one, let me know if that is fine, especially the version

That looks right to me! I will re-review after you address some of the data model comments.

Thanks again!

File: airbyte-api/src/main/openapi/config.yaml [hunk 3]
```diff
@@ -1466,6 +1466,12 @@ components:
       required:
         - webhook
       properties:
+        sendOnSuccess:
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-api/src/main/openapi/config.yaml:1469]: That makes sense, yeah will change

File: airbyte-migration/src/main/java/io/airbyte/migrate/migrations/MigrationV0_30_0.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+  @Override
+  public String getVersion() {
+    return MIGRATION_VERSION;
+  }
+
+  @Override
+  public Map<ResourceId, JsonNode> getOutputSchema() {
+    final Map<ResourceId, JsonNode> outputSchema = new HashMap<>(
+        previousMigration.getOutputSchema());
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-migration/src/main/java/io/airbyte/migrate/migrations/MigrationV0_30_0.java:72]: I guess not. will change

File: airbyte-config/models/src/main/resources/types/SlackNotificationConfiguration.yaml [hunk 5]
```diff
@@ -10,3 +10,9 @@ additionalProperties: false
 properties:
   webhook:
     type: string
+  sendOnSuccess:
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-config/models/src/main/resources/types/SlackNotificationConfiguration.yaml:13]: Will do

**Comment by prasrvenkat** [comment]
> @prasrvenkat thank you for this PR! We really appreciate you digging into so many pieces of our system to put this together. I'm definitely happy to work with you to get this merged!
> 
> Here's the plan.
> 
> * I left a few comments throughout the PR. LMK if anything doesn't make sense. I think there is only one significant change that I am suggesting around the data model.
> * I tried to address some of the questions you left in the PR body below. Again, lmk if anything doesn't make sense!
> * After we iterate one more time, I'll loop in our FE guy to help out with the refreshing issue you mentioned.
> 
> > Ability in front-end to toggle both on/off. I am not that familiar with frontend in general, but tried to get a working solution.
> > The toggle is not auto-save, its tied to the current formik form so reflects only after save changes.
> 
> @Jamakase can you take a look at the FE change in this PR and leave any feedback please?
> 
> > The call site for successful notification is in JobSubmitter after finishing something that is a success. Let me know if this not the right place. Ideally, I thought notification will be downstream option after tracking but there was a notifier already available there so re-used it.
> 
> I agree with your assessment that this is not the ideal place. We want to change it so that the notifier happens from within a Temporal Activity. That being said, I think what you have done for now is a good compromise and we can move the call site in the future. So let's stick with what you have!
> 
> > Migration test was failing so I had to add one, let me know if that is fine, especially the version
> 
> That looks right to me! I will re-review after you address some of the data model comments.
> 
> Thanks again!

Thanks for detailed review, appreciate it. I reworked based on your comment. Let me know what you think @cgardens.

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 6]
```diff
@@ -150,6 +173,32 @@ const WebHookForm: React.FC<WebHookFormProps> = ({
...
             </Message>
           ) : null}
+          <InputRow>
+            <Cell flex={1}>
+              <Field name="sendOnFailure">
+                {({ field }: FieldProps<string>) => (
+                  <LabeledToggle
+                    {...field}
+                    name="sendOnFailure"
+                    checked={webhook.sendOnFailure}
```
[the remark this answers is not in the document]
Comment (Jamakase) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx:183]: Probably that's what you meant in your comment here

> The toggle is not auto-save, its tied to the current formik form so reflects only after save changes.

I think it should be `field.value`. Also could you please change field value to type to boolean above?

**Review by cgardens — APPROVED** [review]
@prasrvenkat thank you! the changes you made on the backend look good to me.

Once you are done iterating on @Jamakase 's feedback on the frontend part, I will help you merge this.

**Comment by Jamakase** [comment]
My overall comment is next: is it actually a good idea to extend notifications interface for every new state we want to add? Is it possible that in the future, we may want to add more types of events on which we notify like `onStart` or update with some `onProgress`? Isn't it better to have generic method that will handle them as `boolean notify(Job job, EventType type)`

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 7]
```diff
@@ -97,12 +111,21 @@ const WebHookForm: React.FC<WebHookFormProps> = ({
...
     <Formik
-      initialValues={{ webhook: notificationUrl }}
+      initialValues={{
+        webhook: webhook.notificationUrl,
+        sendOnSuccess: webhook.sendOnSuccess,
+        sendOnFailure: webhook.sendOnFailure,
+      }}
       validateOnBlur={true}
-      validateOnChange={false}
+      validateOnChange={true}
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: No idea, reverted

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 8]
```diff
@@ -97,12 +111,21 @@ const WebHookForm: React.FC<WebHookFormProps> = ({
...
       validationSchema={webhookValidationSchema}
-      onSubmit={async (values) => {
-        if (notificationUrl && notificationUrl === values.webhook) {
+      onSubmit={async (values: any) => {
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: done

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: Was working with that too, but changed to what you were recommending

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 9]
```diff
@@ -97,12 +111,21 @@ const WebHookForm: React.FC<WebHookFormProps> = ({
...
+        if (
+          webhook &&
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: Used equal

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 10]
```diff
@@ -44,18 +44,32 @@ const Error = styled(Success)`
 
 const webhookValidationSchema = yup.object().shape({
   webhook: yup.string().url("form.url.error"),
+  sendOnSuccess: yup.boolean(),
+  sendOnFailure: yup.boolean(),
 });
 
 type WebHookFormProps = {
-  notificationUrl: string;
+  webhook: {
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: I created WebhookPayload and using it in places. Let me know if the place its defined is fine and name ofcourse

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/NotificationPage.tsx [hunk 11]
```diff
@@ -51,14 +55,26 @@ const NotificationPage: React.FC = () => {
...
+    sendOnFailure: boolean;
+  }) => {
+    await testWebhook(data.webhook, data.sendOnSuccess, data.sendOnFailure);
   };
 
   const initialWebhookUrl =
     workspace.notifications && workspace.notifications.length
       ? workspace.notifications[0].slackConfiguration.webhook
       : "";
+  const initialSendOnSuccess =
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/NotificationPage.tsx]: I don't know what this does, but using it and seems to work..

File: airbyte-webapp/src/components/hooks/services/useWorkspace.tsx [hunk 12]
```diff
@@ -96,18 +104,28 @@ const useWorkspace = (): {
       }
     );
 
-  const testWebhook = async (webhook: string) =>
+  const testWebhook = async (
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/components/hooks/services/useWorkspace.tsx]: Using the new type so just one prop there

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/NotificationPage.tsx [hunk 13]
```diff
@@ -51,14 +55,26 @@ const NotificationPage: React.FC = () => {
...
+    workspace.notifications && workspace.notifications.length
+      ? workspace.notifications[0].sendOnSuccess
+      : false;
+  const initialSendOnFailure =
+    workspace.notifications && workspace.notifications.length
+      ? workspace.notifications[0].sendOnFailure
+      : false;
```
[the remark this answers is not in the document]
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/NotificationPage.tsx]: removed this.

**Comment by prasrvenkat** [comment]
> My overall comment is next: is it actually a good idea to extend notifications interface for every new state we want to add? Is it possible that in the future, we may want to add more types of events on which we notify like `onStart` or update with some `onProgress`? Isn't it better to have generic method that will handle them as `boolean notify(Job job, EventType type)`

If you are talking about toggles then we need individual states if we want to allow that level of control for all such states in notifications.
If you are talking about backend interface/methods, I was just trying to build on top of what is already there. Happy to rework to make it generic if you want. Let me know. cc @cgardens (_But that refactoring might make more sense once we actually have such well-defined EventTypes defined already which afaik is not there yet?_)

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 14]
```diff
@@ -176,23 +158,35 @@ const WebHookForm: React.FC<WebHookFormProps> = ({
...
-                {({ field }: FieldProps<string>) => (
+                {({ field }: FieldProps<boolean>) => (
                   <LabeledToggle
-                    {...field}
-                    name="sendOnFailure"
-                    checked={webhook.sendOnFailure}
+                    name={field.name}
+                    checked={field.value}
+                    onChange={(event) => {
+                      onSubmit({
```
Comment (Jamakase) [inline airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx:166]: Actually I do not think we need to send it on every change. Shouldn't we change only when submit button is pressed? If yes, let's just change it to modify formik state

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx [hunk 15]
```diff
@@ -6,6 +6,8 @@ import * as yup from "yup";
 
 import { Label, Input, LoadingButton, LabeledToggle } from "components";
 import { Row, Cell } from "components/SimpleTableComponents";
+import { WebhookPayload } from "../../../../../components/hooks/services/useWorkspace";
```
Comment (Jamakase) [inline airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx:9]: could you please remove relative imports? so they will start from `components` and `utils`

**Review by jamakase — APPROVED** [review]

File: airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx (hunk 14 above)
Comment (prasrvenkat) [reply airbyte-webapp/src/pages/SettingsPage/pages/NotificationPage/components/WebHookForm.tsx]: Ok, done.

**Comment by marcosmarxm** [comment]
@prasrvenkat could you resolve the conflicts? After that we can start the process of merging this! thanks a lot

**Comment by prasrvenkat** [comment]
> @prasrvenkat could you resolve the conflicts? After that we can start the process of merging this! thanks a lot

@marcosmarxm Done, let me know if anything else needs to be fixed. Thanks

**Review by cgardens — APPROVED** [review]
@prasrvenkat Awesome! This looks like it is ready to go. I am running the acceptance tests on this branch now. Once they are done I will merge.

Thanks for your patience. I was on vacation last week which is why review was postponed.

**Comment by prasrvenkat** [comment]
> @prasrvenkat Awesome! This looks like it is ready to go. I am running the acceptance tests on this branch now. Once they are done I will merge.
> 
> Thanks for your patience. I was on vacation last week which is why review was postponed.

Sounds good, thanks!

**Comment by michel-tricot** [comment]
Thank you @prasrvenkat. This is a great new feature!

**Comment by jrhizor** [comment]
🎉

**Comment by tuliren** [comment]
This is a file based migration, which will mess up our database migration. It needs to be replaced with a Flyway migration.

**Comment by sheshan-doye-konvergeai** [comment]
Is there a way I can update the webhook config programatically ( without going to setting and updating manually) ?

**Comment by prasrvenkat** [comment]
> Is there a way I can update the webhook config programatically ( without going to setting and updating manually) ?

It is part of the workspace so you can check that endpoint in APIs. I don't think a specific endpoint exists just for updating notification afaik.

**Comment by sheshan-doye-konvergeai** [comment]
the web hook doesn't work if endpoint is localhost or 0.0.0.0. it works if i set up my service with tools like ngrok.
but if airbyte is locally deployed it should also detect local endpoints. is it possible ?

**Pull request merged**